examples/diabetes_example/assets/functions/federated_analytics_functions.py

```diff
@@ -0,0 +1,73 @@
+import json
+from collections import defaultdict
+
+import pickle
+
+import numpy as np
+import pandas as pd
+import substratools as tools
+
+
+@tools.register
+def local_step1(inputs, outputs, task_properties):
+    df = inputs["datasamples"]
+    states = {
+        "n_samples": len(df),
+        "means": df.select_dtypes(include=np.number).sum().to_dict(),
+        "counts": {name: series.value_counts().to_dict() for name, series in df.select_dtypes(include='category').items()}
+    }
+    save_states(states, outputs["local_analytics_first_moments"])
+
+
+@tools.register
+def local_step2(inputs, outputs, task_properties):
+    df = inputs["datasamples"]
+    shared_states = load_states(inputs["shared_states"])
+    means = pd.Series(shared_states["means"])
+    states = {
+        "n_samples": len(df),
+        "std": np.power(df.select_dtypes(include=np.number) - means, 2).sum(),
+    }
+    save_states(states, outputs["local_analytics_second_moments"])
+
+
+@tools.register
+def aggregate(inputs, outputs, task_properties):
+    shared_states = [load_states(path) for path in inputs["local_analytics_list"]]
+
+    total_len = 0
+    for state in shared_states:
+        total_len += state["n_samples"]
+
+    aggregated_values = defaultdict(lambda: defaultdict(float))
+    for state in shared_states:
+        for moment, value in state.items():
+            if moment != 'n_samples':
+                for k, v in value.items():
+                    try:
+                        aggregated_values[moment][k] += v / total_len
+                    except TypeError:  # v is not a numeric value, but a dict
```

Comment:
In which case do we have a dict as value ?

Reply:
For categorical features, the value `v` contains a dict of counts for each modality. 

We discussed yesterday that it might be better to rewrite this function in a way that is less generic, but more readable, should I do that? 

Else, I will at least add comments to be more explicit about the data structure.

Reply:
I'm just not convince by the "try except TypeError". What about smthing like:

```
    if isinstance(v, dict):
        if not aggregated_values[moment][k]:
            aggregated_values[moment][k] = defaultdict(float)
        for kk, vv in v.items():
            aggregated_values[moment][k][kk] += vv / total_len
    else:
        aggregated_values[moment][k] += v / total_len
```

I remember you saying that we needed to avoir "isinstance" but I forgot why 😇

Reply:
`isinstance` is making a narrow type check, and it contradicts the idea of duck typing that is usually favored by Python (in short, it doesn't matter if the object is a `dict` or not, as long as it implements a way to get values with the operator `[]`, even without subclassing the `dict` class). 

That's why the `try: ... except` syntax is usually preferred, but I agree that in this particular case, the exception caught is not particularly specific nor explicit. 

So I can stop being pedantic and use a `isinstance(v, dict)` 🙃

Reply:
Ah yes ok thanks for the re-explanation ! 
But yes, in that case where we want to prioritise explicity, I think the usage of `isinstance` might ease the readability of the example 👼